### PR TITLE
Implements the web platform standard reportError API

### PR DESCRIPTION
**Pull request by jasnell**
Refs: https://developer.mozilla.org/en-US/docs/Web/API/reportError

/cc @irvinebroque

### CODE REVIEW

File: samples/helloworld_esm/worker.js [hunk 1]
```diff
@@ -2,8 +2,17 @@
...
+  console.log('reported error:', event.error);
+//  event.preventDefault();
+});
+
+reportError(new DOMException('boom'));
+
 export default {
   async fetch(req, env) {
+    reportError(new DOMException('boom'));
+
```
Comment (jasnell) [inline samples/helloworld_esm/worker.js:15]: The changes in this file are temporary... being used for testing... proper tests and sample files will be added.

File: src/workerd/io/worker.c++ [hunk 2]
```diff
@@ -1025,6 +1025,12 @@ Worker::Isolate::Isolate(kj::Own<Api> apiParam,
         }
         KJ_LOG(INFO, "console warning", message);
       });
+      lock->setErrorReporterCallback([this](jsg::Lock& js, kj::StringPtr message) {
+        if (impl->inspector != kj::none) {
+          logMessage(js, static_cast<uint16_t>(cdp::LogType::ERROR), message);
+        }
+        KJ_LOG(INFO, "unhandled error", message);
```
Comment (jasnell) [inline src/workerd/io/worker.c++:1032]: Need to think about this a bit. When the error is reported to the inspector, the stack is properly reported. But the stack is dropped from the `KJ_LOG` line. Is that ok? Probably not. what else can we do here?

**Comment by irvinebroque** [comment]
If I call `reportError` — what happens here with [Tail Workers](https://developers.cloudflare.com/workers/observability/logging/tail-workers/)? Will the error will show up in the `exceptions` field or the `logs` field?

https://github.com/wintercg/proposal-common-minimum-api/issues/41
https://github.com/nodejs/node/issues/38947

**Comment by jasnell** [comment]
> ... what happens here with Tail Workers?

I think that still needs to be determined. The reporting is the same as the existing `logUncaughtException(...)` but I don't believe that ends up directly in the tail worker's `exceptions` field. That'll be something we want to determine, I think.

**Comment by irvinebroque** [comment]
Yeah we should figure out before merging any change, since would be weird if this ended up showing as logs. @zebp and @maxwellpeterson may know and can point to internal tail worker code

**Comment by jasnell** [comment]
@irvinebroque ... ok, the PR has been updated to publish the reported error to the worker tracer as an exception. Unfortunately we can't actually have a test to verify it in workerd since workerd does not initialize the worker tracer at all. Manually verified in an internal test, however, that the reported error does, in fact, show up in the worker tracer as an exception, so I think we're good on that front.

**Review by irvinebroque — COMMENTED** [review]
Supportive of landing this

**Review by harrishancock — APPROVED** [review]

**Pull request merged**